crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -111,6 +132,43 @@ struct TypeInferenceBuilder<'db> {
     types: TypeInference<'db>,
 }
 
+/// A builder is used by creating it with `::new`, and then calling `.finish()` on it, which
+/// returns the resulting [`TypeInference`].
+///
+/// There are a few different kinds of methods in the type inference builder, and the naming
+/// distinctions are a bit subtle.
+///
+/// The `finish` method calls `infer_region`, which delegates to one of `infer_region_scope`,
+/// `infer_region_definition`, or `infer_region_expression`, depending which kind of
+/// [`InferenceRegion`] we are inferring types for.
+///
+/// Scope inference starts with the scope body, walking all statements and expressions and
+/// recording the types of each expression in the [`TypeInference`] result. Most of the methods
+/// here (with names like `infer_*_statement` or `infer_*_expression` or some other node kind) take
+/// a single AST node and are called as part of this AST visit.
+///
+/// When the visit encounters a node which creates a [`Definition`], we look up the definition in
+/// the semantic index and call the `infer_definition_types` query on it, which creates another
+/// [`TypeInferenceBuilder`] just for that definition, and we merge the returned `TypeInference`
```

Comment:
```suggestion
/// [`TypeInferenceBuilder`] just for that definition, and we merge the returned [`TypeInference`]
```

Sorry for being pedantic. I don't find it that important that these are links. My main motivation is that some IDEs are capable of updating docs when renaming a symbol (and `cargo doc` will fail if we missed renaming a reference in the documentation)

Reply:
> cargo doc will fail if we missed renaming a reference in the documentation

That seems like a good enough motivation to me! But we don't get that benefit on any doc comment on a private item (and this entire module is private), unless we fix https://github.com/astral-sh/ruff/issues/12372 first.